### DISCUSSION TITLE
Differences between react mode and aliases + @tangible/window

### DISCUSSION THREAD

**Discussion by nicolas-jaussaud** [TangibleInc/roller#2] [General]
Hi @eliot-akira!

I have a question regarding the "React mode" argument of the roller (as long as it's not really an issue, I thought I could use this "discussions" feature for it)
 
I'm currently testing a library with the roller ([lexical](https://github.com/facebook/lexical/)), with a setup similar to this one:
```js
module.exports = {
  build: [
    {
      src: 'index.jsx',
      dest: 'build.js',
      alias: {
        'react': '@tangible/window/wp-element',
        'react-dom': '@tangible/window/wp-element'
      }
    }
  ]
}
```
index.jsx being something like this:
```js
import { render } from 'react'
import { LexicalComposer } from '@lexical/react/LexicalComposer'

const div = document.createElement('div')
document.querySelector('body').appendChild(div)

render(
    <div>
        <LexicalComposer initialConfig={{
            namespace: 'tangible'
        }}>
        </LexicalComposer>
    </div>
, div)
```

It works well, but before I didn't use[ @tangible/window](github.com/tangibleInc/tangible-window) and had this config instead:
```js
module.exports = {
  build: [
    {
      src: 'index.jsx',
      dest: 'build.js',
      react: 'wp'
    }
  ]
}
```
Which wasn't working and gave me the following error:
```
Warning: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
```
For what I understood, it was because a [different version of React](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react) was imported by lexical, instead of wp-element

I'm not sure to understand why it works differently when using an alias with `@tangible/window`, and until now I was assuming that it was the same thing than using `react: 'wp'`

I was wondering what the differences between the two approaches are, and which method should be used in which cases

**Comment by eliot-akira** [comment]
Hi @nicolas-jaussaud, thanks for the question.

You're right that `react: 'wp'` is supposed to work the same as `@tangible/window`. It's meant to be a simpler solution that doesn't require an additional NPM package.

The `@tangible/window` package was a workaround for aliasing imports to global variables like `wp.element`, `Tangible.Preact`, and `jQuery`.

The new solution is to use [`rollup-plugin-external-globals`](https://github.com/eight04/rollup-plugin-external-globals). In the task config, it's the `importToGlobal` option, which gets passed to this plugin.

Internally, React mode `wp` sets the `importToGlobal` option for `react` and `react-dom` to reference `wp.element`.

https://github.com/TangibleInc/tangible-roller/blob/37067684eb1a63da231ffada567e09a91e1d7259/task/js.js#L130-L136

From your description, it sounds like this is not being applied to the `lexical` package in `node_modules`.

..Oh, maybe you're importing the compiled/built library files of `lexical`, which uses the `require` method (called CommonJS) to import React. And it's possible this Rollup plugin only works with `export/import` (called ES Modules). (For explanation of these import syntaxes: [CommonJS vs. ES modules in Node.js](https://blog.logrocket.com/commonjs-vs-es-modules-node-js/).) Mmm, but when I implemented the React modes, I'm pretty sure I checked that it works with compiled packages in `node_modules`..

So I would say this is a bug in the React mode `wp`. I'll try to recreate the issue locally and see if I can find a solution.

(Ah, cool - I can "Create issue from discussion" in the GitHub sidebar menu. Issue #3 )

**Comment by eliot-akira** [comment]
Whew, it took a couple hours of digging deep into the Rollup plugins, but I believe I solved it in commit https://github.com/TangibleInc/tangible-roller/commit/d4a508d8b6e0c2b35a855810e0fd59e2b365c846.

The cause of the issue was that for `require` statements, the CommonJS plugin transforms the module names  into absolute paths, for example, `react` into `.../node_modules/react/cjs/react.production.min.js`. (This only happens if regular `react` is installed in `node_modules`. Otherwise, the module name is left as is.) That's why the External Globals plugin didn't recognize it in order to transform it into a global variable, in this case `wp.element`.

The solution was to pass to the latter plugin a function that matches the module name as is (for `import` statements) and as absolute paths (for `require` statements).

It's been published as version `1.4.4`. Could you run `npm update` in your local project and see if the React mode `wp` works now?

**Comment by nicolas-jaussaud** [comment]
Thank you for the quick update!

I tested again and while components imported from the lexical library are now using wp.element, it seems that there is an issue in the way it's imported

Here is the error I got after updating and building a new file using the `react: 'wp'` option (I tested on the same `index.jsx` than in my first message):
```
Uncaught TypeError: window.wp.element.__require is not a function
```

In the built file, I have indeed multiple occurrences of something like:
```js
var react = wp.element.__require();
```
If I manually replace every occurence of `wp.element.__require()` by `wp.element`, the script works as expected

**Reply by eliot-akira** [reply]
OK, I tracked it down to this part of the CommonJS plugin, which wraps a module with `__require` function in some cases.

https://github.com/rollup/plugins/blob/45503a639b29619648f4394a8e003c91edff3601/packages/commonjs/src/generate-imports.js#L197-L203

In commit 4e9aaf2, I updated the module name matching function for the externalGlobals plugin to detect such wrapped modules and replace it with a new function that returns the global variable.

With my local minimal test, I wasn't able to fully reproduce the issue (only the __require part but not the code that calls it). Could you update to Roller 1.4.5 and see if that works?

---

This is an admittedly hacky workaround, as it depends on the internals of the CommonJS plugin. The real issue is that Rollup itself has [an option called `globals`](https://rollupjs.org/guide/en/#outputglobals) that's supposed to replace imports with globals, but it only works in limited cases - that's why someone had to create [rollup-plugin-external-globals](https://github.com/eight04/rollup-plugin-external-globals). It also doesn't work with the CommonJS plugin, apparently, as the plugin transforms require statements into imports with absolute paths regardless of the `globals` option.

The ideal solution would be for the CommonJS plugin to apply the `globals` option as expected. But the plugin is complex, and the maintainers of Rollup require a reproducible example accompanying an issue report - which will take time to prepare.

**Reply by eliot-akira** [reply]
Ah, actually - I think I found a much better solution.

In commit 123e409, I added a new Rollup [option called `external`](https://rollupjs.org/guide/en/#external) so that any imported module aliased to global variable will not be transformed by the CommonJS plugin. And I removed the whole workaround with absolute module path and require function wrapper.

Please try Roller 1.4.6. Hopefully that works, because this would be a clean and satisfying solution. If not, I'll restore the workaround as a fallback.

**Reply by nicolas-jaussaud** [reply]
Hi!

Thank you for your help, updating to 1.4.6 indeed fixes the issue!

Unfortunately, it seems to cause a different error with another import from another library, which was working previously ([react-transition-group](https://github.com/reactjs/react-transition-group))

In the files from their npm module, react is required like this:
```js

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { 
    return obj && obj.__esModule ? obj : { default: obj }; 
}

function (_React$Component) {
   function ComponentName(props, context) {
    // ...
   }
}(_react.default.Component);

``` 
In our built file, the `_react` variable becomes something like this:
```js
var _react = _interopRequireDefault(_global_wp_element$1);
```

The issue is that while `_global_wp_element$1.__esModule` is true, `_global_wp_element$1.default` does not exist

But as long as the `_interopRequireDefault$4` relies on the `__esModule` attribute to set or not a `default` value, it will not be set and will throws an error when reaching the `_react$3.default.Component` part:

```
Uncaught TypeError: _react$3.default is undefined
```

I guess somehow the roller was setting this default value before?

(I attached a small wordpress plugin that replicates the issue)
[tangible-test.zip](https://github.com/TangibleInc/tangible-roller/files/9876524/tangible-test.zip)

**Reply by eliot-akira** [reply]
Oh my, what a rabbit hole this is turning out to be, haha.

The `__esModule` wrapper looks like something that the CommonJS plugin should be processing. OK, so I'll revert to the workaround in the previous version of Roller, which lets the aliased modules go through CommonJS (and then overwrite them as globals).

I appreciate the example zip, that will help me confirm when I actually get it working.

---

The real underlying cause of issue is that [`rollup-plugin-external-globals`](https://github.com/eight04/rollup-plugin-external-globals) is not working well with [`@rollup/plugin-commonjs`](https://github.com/rollup/plugins/tree/master/packages/commonjs). (And also that Rollup's option `globals` is not working as expected.)

It's almost too much for me to try to solve their compatibility issue - these JavaScript transpilers are crazy complex. When I have more time, I will see if I can prepare an issue report and hopefully the plugin author can solve it properly.

But in the meantime, I think I'm close to finding an acceptable temporary solution.

**Reply by eliot-akira** [reply]
Commit 055be89 should do it. Please try Roller version `1.4.7` by running `npm update`.

**Reply by nicolas-jaussaud** [reply]
Thanks a lot!
